### ISSUE TITLE
Make "--feature-gates" flag component-related

### ISSUE THREAD

**Issue by xiangpengzhao** [kubernetes/kubernetes#46822]
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see https://kubernetes.io/docs/tasks/debug-application-cluster/troubleshooting/.):

**Note:** Please file issues for subcomponents under the appropriate repo

| Component | Repo                                                               |
| --------- | ------------------------------------------------------------------ |
| kubectl   | [kubernetes/kubectl](https://github.com/kubernetes/kubectl/issues/new) |
| kubeadm   | [kubernetes/kubeadm](https://github.com/kubernetes/kubeadm/issues/new) |

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):
Enhancement.

**Kubernetes version** (use `kubectl version`):
HEAD

**Environment**:
- **Cloud provider or hardware configuration**:
- **OS** (e.g. from /etc/os-release):
- **Kernel** (e.g. `uname -a`):
- **Install tools**:
- **Others**:

**What happened**:
We have `--feature-gates` flag in help info of every main components. And all of them have the same options listed.

```
Accelerators=true|false (ALPHA - default=false)
AdvancedAuditing=true|false (ALPHA - default=false)
AffinityInAnnotations=true|false (ALPHA - default=false)
AllAlpha=true|false (ALPHA - default=false)
AllowExtTrafficLocalEndpoints=true|false (default=true)
AppArmor=true|false (BETA - default=true)
DynamicKubeletConfig=true|false (ALPHA - default=false)
DynamicVolumeProvisioning=true|false (ALPHA - default=true)
ExperimentalCriticalPodAnnotation=true|false (ALPHA - default=false)
ExperimentalHostUserNamespaceDefaulting=true|false (BETA - default=false)
PersistentLocalVolumes=true|false (ALPHA - default=false)
StreamingProxyRedirects=true|false (BETA - default=true)
TaintBasedEvictions=true|false (ALPHA - default=false)
```

When users want to switch a feature gate on, it's hard to know which components need to enable the flag. Take `AffinityInAnnotations` for example, users might know they should enable this flag for apiserver and scheduler. But how about controller-manager and kubelet? They might need or not. For kube-proxy, it obviously doesn't need to enable this flag.

**What you expected to happen**:
Only show the options for components which need them. It would be good to have a documentation to guide users.

**How to reproduce it** (as minimally and precisely as possible):

**Anything else we need to know**:

/sig api-machinery

**Comment by luxas** [comment]
Absolutely agree, I think this would be a good improvement for v1.8

cc @jamiehannaford I guess your PR can address this issue? It will solve the problem for kubeadm at the same time...

**Comment by jamiehannaford** [comment]
Yeah, sure. I'll work on something and submit it sometime this week. 

@xiangpengzhao If you'd like to review when it's ready that'd also be great.

**Comment by xiangpengzhao** [comment]
@jamiehannaford Sure. I'm look forward to your PR! Please ping me when it's ready. Thanks for your work!

**Comment by jamiehannaford** [comment]
@xiangpengzhao I think this is going to take a bit of work to do, because lots of the current codebase is coupled to the current global state. I'm seeing a lot of flagset errors when I try to decouple it. Unfortunately I don't have much bandwidth at the moment to dedicate to this, but might be able to circle back in a few weeks.

**Comment by xiangpengzhao** [comment]
@jamiehannaford thanks for your work on this! I really appreciate it! Hope we will find an ideal solution:)

**Comment by josrom** [comment]
Have same issue i need the functionality to be able to delete pods from a node that has died:
Version 1.8 introduces an alpha feature that automatically creates taints that represent conditions. To enable this behavior, pass an additional feature gate flag --feature-gates=...,TaintNodesByCondition=true to the API server, controller manager, and scheduler. When TaintNodesByCondition is enabled, the scheduler ignores conditions when considering a Node; instead it looks at the Node’s taints and a Pod’s tolerations.
source: https://kubernetes.io/docs/concepts/architecture/nodes/

**Comment by brightzheng100** [comment]
I'm also very confused while thinking which component should be updated to enable (or disable) a feature.

Let’s take `RotateKubeletClientCertificate` as an example.
From the command line reference doc, it looks like we have to update any of below components:
- cloud-controller-manager
- kube-apiserver
- kube-controller-manager
- kube-proxy
- kube-scheduler

But I'm not sure, frankly speaking, if purely based on the docs.
So for now, maybe a little doc improvement may help by highlighting something like:

| feature | Component A | Component B | Component C | Component D |
| ------------- | :---: | :---: | :---: | :---: |
|  RotateKubeletClientCertificate | | | Y | Y | Y |

**Comment by neolit123** [comment]
there have been recent discussions around doing that, as components might be breaking out of tree - e.g. kube-scheduler.
[1] but the overall idea of having separate set of feature gates for separate components is still seeing discussion:
https://github.com/kubernetes/kubernetes/pull/90111#discussion_r407776222

kubeadm already implements it's own set of feature gates that are only relevant to it.

@brightzheng100 given [1], your idea of documenting a component matrix is a good one.
some feature flags simply don't make sense to some components, so this should be documented.

ideally this should be owned by:
/wg component-standard
/kind design documentation
/remove-sig cluster-lifecycle api-machinery

**Comment by neolit123** [comment]
...but best to have some sort of a way to automate / generate the documentation for the feature gates on each release or it's going to drift.

**Comment by MadhavJivrajani** [comment]
/remove-kind design
/kind feature

`kind/design` will soon be removed from `k/k` in favor of `kind/feature`. Relevant discussion can be found here: https://github.com/kubernetes/community/issues/5641
